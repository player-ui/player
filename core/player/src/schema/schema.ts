import { SyncWaterfallHook } from "tapable-ts";
import type { Schema as SchemaType, Formatting } from "@player-ui/types";

import type { BindingInstance } from "../binding";
import type { ValidationProvider, ValidationObject } from "../validator";
import type { FormatDefinition, FormatOptions, FormatType } from "./types";

/** A function that returns itself */
const identify = (val: any) => val;

/** Expand the authored schema into a set of paths -> DataTypes */
export function parse(
  schema: SchemaType.Schema,
): Map<string, SchemaType.DataTypes> {
  const expandedPaths = new Map<string, SchemaType.DataTypes>();

  if (!schema.ROOT) {
    return expandedPaths;
  }

  const parseQueue: Array<{
    /** The node to process */
    node: SchemaType.Node;

    /** The path in the data-model this node represents */
    path: Array<string>;

    /** A set of visited DataTypes to prevent loops */
    visited: Set<string>;
  }> = [{ node: schema.ROOT, path: [], visited: new Set() }];

  while (parseQueue.length > 0) {
    const next = parseQueue.shift();

    if (!next) {
      break;
    }

    const { node, path, visited } = next;

    Object.entries(node).forEach(([prop, type]) => {
      const nestedPath = [...path, prop];

      const nestedPathStr = nestedPath.join(".");

      if (expandedPaths.has(nestedPathStr)) {
        // We've gone in a loop. Panic
        throw new Error(
          "Path has already been processed. There's either a loop somewhere or a bug",
        );
      }

      if (visited.has(type.type)) {
        throw new Error(
          `Path already contained type: ${type.type}. This likely indicates a loop in the schema`,
        );
      }

      expandedPaths.set(nestedPathStr, type);

      if (type.isArray) {
        nestedPath.push("[]");
      }

      if (type.isRecord) {
        nestedPath.push("{}");
      }

      if (type.type && schema[type.type]) {
        parseQueue.push({
          path: nestedPath,
          node: schema[type.type],
          visited: new Set([...visited, type.type]),
        });
      }
    });
  }

  return expandedPaths;
}

/**
 * The Schema is the central hub for all data invariants, and metaData associated with the data-model itself
 * Outside of the types defined in the JSON payload, it doesn't manage or keep any state.
 * It simply servers as an orchestrator for other modules to interface w/ the schema.
 */
export class SchemaController implements ValidationProvider {
  private formatters: Map<string, FormatType<any, any, FormatOptions>> =
    new Map();

  private types: Map<string, SchemaType.DataType<any>> = new Map();
  public readonly schema: Map<string, SchemaType.DataTypes> = new Map();

  private bindingSchemaNormalizedCache: Map<BindingInstance, string> =
    new Map();

  public readonly hooks = {
    resolveTypeForBinding: new SyncWaterfallHook<
      [SchemaType.DataTypes | undefined, BindingInstance]
    >(),
  };

  constructor(schema?: SchemaType.Schema) {
    this.schema = schema ? parse(schema) : new Map();
  }

  public addFormatters(fns: Array<FormatType<any, any, FormatOptions>>) {
    fns.forEach((def) => {
      this.formatters.set(def.name, def);
    });
  }

  public addDataTypes(types: Array<SchemaType.DataType<any>>) {
    types.forEach((t) => {
      this.types.set(t.type, t);
    });
  }

  getValidationsForBinding(
    binding: BindingInstance,
  ): Array<ValidationObject> | undefined {
    const typeDef = this.getApparentType(binding);

    if (!typeDef?.validation?.length) {
      return undefined;
    }

    // Set the defaults for schema-level validations
    return typeDef.validation.map((vRef) => ({
      severity: "error",
      trigger: "change",
      ...vRef,
    }));
  }

  private normalizeBinding(binding: BindingInstance): string {
    const cached = this.bindingSchemaNormalizedCache.get(binding);
    if (cached) {
      return cached;
    }

    let bindingArray = binding.asArray();
    let normalized = bindingArray
      .map((p) => (typeof p === "number" ? "[]" : p))
      .join(".");

    if (normalized) {
      this.bindingSchemaNormalizedCache.set(binding, normalized);
      bindingArray = normalized.split(".");
    }

    bindingArray.forEach((item) => {
      const recordBinding = bindingArray
        .map((p) => (p === item ? "{}" : p))
        .join(".");

      if (this.schema.get(recordBinding)) {
        this.bindingSchemaNormalizedCache.set(binding, recordBinding);
        bindingArray = recordBinding.split(".");
        normalized = recordBinding;
      }
    });

    return normalized;
  }

  public getType(binding: BindingInstance): SchemaType.DataTypes | undefined {
    return this.hooks.resolveTypeForBinding.call(
      this.schema.get(this.normalizeBinding(binding)),
      binding,
    );
  }

  public getApparentType(
    binding: BindingInstance,
  ): SchemaType.DataTypes | undefined {
    const schemaType = this.getType(binding);

    if (schemaType === undefined) {
      return undefined;
    }

    const baseType = this.getTypeDefinition(schemaType?.type);

    if (baseType === undefined) {
      return schemaType;
    }

    return {
      ...baseType,
      ...schemaType,
      validation: [
        ...(schemaType.validation ?? []),
        ...(baseType.validation ?? []),
      ],
    };
  }

  public getTypeDefinition(dataType: string) {
    return this.types.get(dataType);
  }

  public getFormatterForType(
    formatReference: Formatting.Reference,
  ): FormatDefinition<unknown, unknown> | undefined {
    const { type: formatType, ...options } = formatReference;

    const formatter = this.formatters.get(formatType);

    if (!formatter) {
      return;
    }

    return {
      format: formatter.format
        ? (val) => formatter.format?.(val, options)
        : identify,
      deformat: formatter.deformat
        ? (val) => formatter.deformat?.(val, options)
        : identify,
    };
  }

  /**
   * Given a binding, fetch a function that's responsible for formatting, and/or de-formatting the data
   * If no formatter is registered, it will return undefined
   */
  public getFormatter(
    binding: BindingInstance,
  ): FormatDefinition<unknown, unknown> | undefined {
    const type = this.getApparentType(binding);

    if (!type?.format) {
      return undefined;
    }

    return this.getFormatterForType(type.format);
  }
}
