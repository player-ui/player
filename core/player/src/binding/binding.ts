import { getBindingSegments } from "./utils";

export interface BindingParserOptions {
  /** Get the value for a specific binding */
  get: (binding: BindingInstance) => any;

  /**
   * Set the values for bindings.
   * This is used when the query syntax needs to modify an object
   */
  set: (transaction: Array<[BindingInstance, any]>) => void;

  /**
   * Get the result of evaluating an expression
   */
  evaluate: (exp: string) => any;

  /**
   * Without readOnly, if a binding such as this is used: arr[key='does not exist'],
   * then an object with that key will be created.
   * This is done to make assignment such as arr[key='abc'].val = 'foo' work smoothly.
   * Setting readOnly to true will prevent this behavior, avoiding unintended data changes.
   */
  readOnly?: boolean;
}

export type Getter = (path: BindingInstance) => any;

export type RawBindingSegment = number | string;
export type RawBinding = string | RawBindingSegment[];
export type BindingLike = RawBinding | BindingInstance;
export type BindingFactory = (
  raw: RawBinding,
  options?: Partial<BindingParserOptions>,
) => BindingInstance;

/**
 * A path in the data model
 */
export class BindingInstance {
  private split: RawBindingSegment[];
  private joined: string;
  private factory: BindingFactory;

  constructor(
    raw: RawBinding,
    factory = (rawBinding: RawBinding) => new BindingInstance(rawBinding),
  ) {
    const split = Array.isArray(raw) ? raw : raw.split(".");
    this.split = split.map((segment) => {
      if (typeof segment === "number") {
        return segment;
      }

      const tryNum = Number(segment);
      return isNaN(tryNum) ? segment : tryNum;
    });
    Object.freeze(this.split);
    this.joined = this.split.join(".");
    this.factory = factory;
  }

  asArray(): RawBindingSegment[] {
    return this.split;
  }

  asString(): string {
    return this.joined;
  }

  /**
   * Check to see if the given binding is a sub-path of the current one
   */
  contains(binding: BindingInstance): boolean {
    // need to account for partial key matches
    // [foo, bar] !== [foo, ba]
    const bindingAsArray = binding.asArray();

    if (bindingAsArray.length < this.split.length) {
      return false;
    }

    // Check every overlapping index to make sure they're the same
    // Intentionally use a for loop for speeeed
    for (let i = 0; i < this.split.length; i++) {
      if (this.split[i] !== bindingAsArray[i]) {
        return false;
      }
    }

    return true;
  }

  relative(binding: BindingInstance): RawBindingSegment[] {
    return this.asArray().slice(binding.asArray().length);
  }

  parent(): BindingInstance {
    return this.factory(this.split.slice(0, -1));
  }

  key(): RawBindingSegment {
    return this.split[this.split.length - 1];
  }

  /**
   * This is a utility method to get a binding that is a descendent of this binding
   *
   * @param relative - The relative path to descend to
   */
  descendent(relative: BindingLike): BindingInstance {
    const descendentSegments = getBindingSegments(relative);

    return this.factory(this.split.concat(descendentSegments));
  }
}
