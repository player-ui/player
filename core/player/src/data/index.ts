export * from './dependency-tracker';
export * from './model';
export * from './noop-model';
export * from './local-model';
