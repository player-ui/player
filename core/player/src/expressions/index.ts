export * from "./evaluator";
export * from "./types";
export * from "./utils";
export * from "./parser";
