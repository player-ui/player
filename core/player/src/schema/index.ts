export * from "./types";
export * from "./schema";
