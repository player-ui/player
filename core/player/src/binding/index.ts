import { SyncBailHook, SyncWaterfallHook } from 'tapable-ts';
import NestedError from 'nested-error-stacks';
import type { ParserResult, AnyNode } from '../binding-grammar';
import {
  // We can swap this with whichever parser we want to use
  parseCustom as parseBinding,
} from '../binding-grammar';
import type { BindingParserOptions, BindingLike } from './binding';
import { BindingInstance } from './binding';
import { isBinding } from './utils';
import type { NormalizedResult, ResolveBindingASTOptions } from './resolver';
import { resolveBindingAST } from './resolver';

export * from './utils';
export * from './binding';

export const SIMPLE_BINDING_REGEX = /^[\w\-@]+(\.[\w\-@]+)*$/;

const DEFAULT_OPTIONS: BindingParserOptions = {
  get: () => {
    throw new Error('Not Implemented');
  },
  set: () => {
    throw new Error('Not Implemented');
  },
  evaluate: () => {
    throw new Error('Not Implemented');
  },
};

/** A parser for creating bindings from a string */
export class BindingParser {
  private cache: Record<string, BindingInstance>;
  private parseCache: Record<string, ParserResult>;
  private parserOptions: BindingParserOptions;

  public hooks = {
    skipOptimization: new SyncBailHook<[string], boolean>(),
    beforeResolveNode: new SyncWaterfallHook<
      [AnyNode, Required<NormalizedResult> & ResolveBindingASTOptions]
    >(),
  };

  constructor(options?: Partial<BindingParserOptions>) {
    this.parserOptions = { ...DEFAULT_OPTIONS, ...options };
    this.cache = {};
    this.parseCache = {};
    this.parse = this.parse.bind(this);
  }

  /**
   * Takes a binding path, parses it, and returns an equivalent, normalized
   * representation of that path.
   */
  private normalizePath(
    path: string,
    resolveOptions: ResolveBindingASTOptions
  ) {
    if (
      path.match(SIMPLE_BINDING_REGEX) &&
      this.hooks.skipOptimization.call(path) !== true
    ) {
      return { path: path.split('.'), updates: undefined } as NormalizedResult;
    }

    const ast = this.parseCache[path] ?? parseBinding(path);
    this.parseCache[path] = ast;

    if (typeof ast !== 'object' || !ast?.status) {
      throw new TypeError(
        `Cannot normalize path "${path}": ${ast?.error ?? 'Unknown Error.'}`
      );
    }

    try {
      return resolveBindingAST(ast.path, resolveOptions, this.hooks);
    } catch (e: any) {
      throw new NestedError(`Cannot resolve binding: ${path}`, e);
    }
  }

  private getBindingForNormalizedResult(
    normalized: NormalizedResult
  ): BindingInstance {
    const normalizedStr = normalized.path.join('.');

    if (this.cache[normalizedStr]) {
      return this.cache[normalizedStr];
    }

    const created = new BindingInstance(
      normalizedStr === '' ? [] : normalized.path,
      this.parse
    );
    this.cache[normalizedStr] = created;

    return created;
  }

  public parse(
    rawBinding: BindingLike,
    overrides: Partial<BindingParserOptions> = {}
  ): BindingInstance {
    if (isBinding(rawBinding)) {
      return rawBinding;
    }

    const options = {
      ...this.parserOptions,
      ...overrides,
    };

    let updates: Record<string, any> = {};

    const joined = Array.isArray(rawBinding)
      ? rawBinding.join('.')
      : String(rawBinding);

    const normalizeConfig: ResolveBindingASTOptions = {
      getValue: (path: Array<string | number>) => {
        const normalized = this.normalizePath(path.join('.'), normalizeConfig);

        return options.get(this.getBindingForNormalizedResult(normalized));
      },
      evaluate: (exp) => {
        return options.evaluate(exp);
      },
      convertToPath: (path: any) => {
        if (path === undefined) {
          throw new Error(
            'Attempted to convert undefined value to binding path'
          );
        }

        if (
          typeof path !== 'string' &&
          typeof path !== 'number' &&
          typeof path !== 'boolean'
        ) {
          throw new Error(
            `Attempting to convert ${typeof path} to a binding path.`
          );
        }

        const normalized = this.normalizePath(String(path), normalizeConfig);

        if (normalized.updates) {
          updates = {
            ...updates,
            ...normalized.updates,
          };
        }

        const joinedNormalizedPath = normalized.path.join('.');

        if (joinedNormalizedPath === '') {
          throw new Error('Nested path resolved to an empty path');
        }

        return joinedNormalizedPath;
      },
    };

    const normalized = this.normalizePath(joined, normalizeConfig);

    if (normalized.updates) {
      updates = {
        ...updates,
        ...normalized.updates,
      };
    }

    const updateKeys = Object.keys(updates);

    if (!options.readOnly && updateKeys.length > 0) {
      const updateTransaction = updateKeys.map<[BindingInstance, any]>(
        (updatedBinding) => [
          this.parse(updatedBinding),
          updates[updatedBinding],
        ]
      );

      options.set(updateTransaction);
    }

    return this.getBindingForNormalizedResult(normalized);
  }
}
