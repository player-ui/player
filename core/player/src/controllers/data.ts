import { SyncHook, SyncWaterfallHook, SyncBailHook } from 'tapable-ts';
import { omit, removeAt } from 'timm';
import { dequal } from 'dequal';
import type { Logger } from '../logger';
import type { BindingParser, BindingLike } from '../binding';
import { BindingInstance } from '../binding';
import type {
  BatchSetTransaction,
  Updates,
  DataModelOptions,
  DataModelWithParser,
  DataPipeline,
  DataModelMiddleware,
} from '../data';
import { PipelinedDataModel, LocalModel } from '../data';
import type { RawSetTransaction } from '../types';

/** The orchestrator for player data */
export class DataController implements DataModelWithParser<DataModelOptions> {
  public hooks = {
    resolve: new SyncWaterfallHook(),
    resolveDataStages: new SyncWaterfallHook<[DataPipeline]>(),

    // On any set or get of an undefined value, redirect the value to be the default
    resolveDefaultValue: new SyncBailHook<[BindingInstance], any>(),

    onDelete: new SyncHook<[any]>(),

    onSet: new SyncHook<[BatchSetTransaction]>(),

    onGet: new SyncHook<[any, any]>(),

    onUpdate: new SyncHook<[Updates, DataModelOptions | undefined]>(),

    format: new SyncWaterfallHook<[any, BindingInstance]>(),

    deformat: new SyncWaterfallHook<[any, BindingInstance]>(),

    serialize: new SyncWaterfallHook<[any]>(),
  };

  private model?: PipelinedDataModel;
  private trash: Set<BindingInstance>;
  private pathResolver: BindingParser;
  private baseMiddleware: Array<DataModelMiddleware>;
  private logger?: Logger;

  constructor(
    model: Record<any, unknown> | undefined,
    options: {
      /** A means of parsing a raw binding to a Binding object */
      pathResolver: BindingParser;

      /** middleware to use. typically for validation */
      middleware?: Array<DataModelMiddleware>;

      /** A logger to use  */
      logger?: Logger;
    }
  ) {
    this.logger = options.logger;
    const middleware = options.middleware || [];
    this.baseMiddleware = [new LocalModel(model), ...middleware];

    this.trash = new Set();
    this.pathResolver = options.pathResolver;
  }

  public getModel(): PipelinedDataModel {
    if (!this.model) {
      const stages = this.hooks.resolveDataStages.call(this.baseMiddleware);
      const model = new PipelinedDataModel();
      model.setMiddleware(stages);
      this.model = model;
    }

    return this.model;
  }

  private resolveDataValue(
    binding: BindingInstance,
    value: any,
    deformat: boolean
  ) {
    if (deformat) {
      return this.hooks.deformat.call(value, binding);
    }

    return value;
  }

  public set(
    transaction: RawSetTransaction,
    options?: DataModelOptions
  ): Updates {
    let normalizedTransaction: BatchSetTransaction = [];

    if (Array.isArray(transaction)) {
      normalizedTransaction = transaction.map(([binding, value]) => {
        const parsed = this.pathResolver.parse(binding);

        return [
          parsed,
          this.resolveDataValue(parsed, value, Boolean(options?.formatted)),
        ];
      }) as BatchSetTransaction;
    } else {
      normalizedTransaction = Object.keys(transaction).map(
        (binding: string) => {
          const parsed = this.pathResolver.parse(binding);
          const val = transaction[binding];

          return [
            parsed,
            this.resolveDataValue(parsed, val, Boolean(options?.formatted)),
          ];
        }
      ) as BatchSetTransaction;
    }

    // Figure out what the base changes being applied are
    const setUpdates = normalizedTransaction.reduce<Updates>(
      (updates, [binding, newVal]) => {
        const oldVal = this.get(binding, { includeInvalid: true });

        const update = {
          binding,
          newValue: newVal,
          oldValue: oldVal,
        };

        if (dequal(oldVal, newVal)) {
          this.logger?.debug(
            `Skipping update for path: ${binding.asString()}. Value was unchanged: ${oldVal}`
          );
        } else {
          updates.push(update);

          this.logger?.debug(
            `Setting path: ${binding.asString()} from: ${oldVal} to: ${newVal}`
          );
        }

        return updates;
      },
      []
    );

    // Get the applied update
    const result = this.getModel().set(normalizedTransaction, options);

    // Add any extra bindings that were effected
    const setUpdateBindings = new Set(setUpdates.map((su) => su.binding));
    result.forEach((tr) => {
      if (
        !setUpdateBindings.has(tr.binding) &&
        (tr.force === true || !dequal(tr.oldValue, tr.newValue))
      ) {
        this.logger?.debug(
          `Path: ${tr.binding.asString()} was changed from: ${
            tr.oldValue
          } to: ${tr.newValue}`
        );
        setUpdates.push(tr);
      }
    });

    this.hooks.onSet.call(normalizedTransaction);

    if (setUpdates.length > 0) {
      this.hooks.onUpdate.call(setUpdates, options);
    }

    return result;
  }

  private resolve(binding: BindingLike, readOnly: boolean): BindingInstance {
    return Array.isArray(binding) || typeof binding === 'string'
      ? this.pathResolver.parse(binding, { readOnly })
      : binding;
  }

  public get(binding: BindingLike, options?: DataModelOptions) {
    const resolved =
      binding instanceof BindingInstance
        ? binding
        : this.resolve(binding, true);
    let result = this.getModel().get(resolved, options);

    if (result === undefined && !options?.ignoreDefaultValue) {
      const defaultVal = this.hooks.resolveDefaultValue.call(resolved);

      if (defaultVal !== result) {
        result = defaultVal;
      }
    }

    if (options?.formatted) {
      result = this.hooks.format.call(result, resolved);
    } else if (options?.formatted === false) {
      result = this.hooks.deformat.call(result, resolved);
    }

    this.hooks.onGet.call(binding, result);

    return result;
  }

  public delete(binding: BindingLike, options?: DataModelOptions) {
    if (
      typeof binding !== 'string' &&
      !Array.isArray(binding) &&
      !(binding instanceof BindingInstance)
    ) {
      throw new Error('Invalid arguments: delete expects a data path (string)');
    }

    const resolved =
      binding instanceof BindingInstance
        ? binding
        : this.resolve(binding, false);

    const parentBinding = resolved.parent();
    const property = resolved.key();
    const parentValue = this.get(parentBinding);

    const existedBeforeDelete =
      typeof parentValue === 'object' &&
      parentValue !== null &&
      Object.prototype.hasOwnProperty.call(parentValue, property);

    this.getModel().delete(resolved, options);

    if (existedBeforeDelete && !this.get(resolved)) {
      this.trash.add(resolved);
    }

    this.hooks.onDelete.call(resolved);
  }

  public serialize(): object {
    return this.hooks.serialize.call(this.get(''));
  }
}
