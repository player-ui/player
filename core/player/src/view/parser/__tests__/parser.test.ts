import { BindingParser } from '../../../binding';
import { LocalModel, withParser } from '../../../data';
import { SchemaController } from '../../../schema';
import { NodeType, Parser } from '../index';
import { SwitchPlugin, ApplicabilityPlugin, TemplatePlugin } from '../..';
import type { Options } from '../../plugins/options';
import { ExpressionEvaluator } from '../../../expressions';
import type { DataModelWithParser } from '../../../data';

const parseBinding = new BindingParser().parse;
describe('generates the correct AST', () => {
  let model: DataModelWithParser;
  let expressionEvaluator: ExpressionEvaluator;
  let options: Options;
  let parser: Parser;

  beforeEach(() => {
    model = withParser(new LocalModel(), parseBinding);
    expressionEvaluator = new ExpressionEvaluator({
      model,
    });
    parser = new Parser();
    options = {
      evaluate: expressionEvaluator.evaluate,
      schema: new SchemaController(),
      data: {
        format: (binding, val) => val,
        formatValue: (val) => val,
        model,
      },
    };
    new TemplatePlugin(options).applyParser(parser);
    new ApplicabilityPlugin().applyParser(parser);
    new SwitchPlugin(options).applyParser(parser);
  });

  test('works with basic objects', () => {
    expect(parser.parseObject({ foo: 'bar' })).toStrictEqual({
      type: NodeType.Value,
      value: {
        foo: 'bar',
      },
    });
  });

  test('works with objects that have symbols', () => {
    const testSymbol = Symbol('foo');
    expect(parser.parseObject({ [testSymbol]: 'bar' })).toStrictEqual({
      type: NodeType.Value,
      value: {
        [testSymbol]: 'bar',
      },
    });
  });

  test('works with applicability things', () => {
    expect(
      parser.parseObject({ foo: 'bar', applicability: '{{baz}}' })
    ).toMatchSnapshot();

    expect(
      parser.parseObject({
        asset: {
          values: [
            {
              applicability: '{{foo}}',
              value: 'foo',
            },
            {
              value: 'bar',
            },
          ],
        },
      })
    ).toMatchSnapshot();

    expect(
      parser.parseObject({
        asset: {
          someProp: {
            applicability: '{{foo}}',
            label: {
              value: 'label',
            },
            description: {
              value: 'description',
            },
          },
        },
      })
    ).toMatchSnapshot();

    expect(
      parser.parseObject({
        asset: {
          someProp: {
            asset: {
              applicability: '{{foo}}',
              type: 'someAsset',
            },
          },
        },
      })
    ).toMatchSnapshot();
  });

  test('parses an object', () => {
    expect(parser.parseObject({ asset: { type: 'bar' } })).toMatchSnapshot();
  });

  test('parses an exp array', () => {
    expect(
      parser.parseObject(
        {
          id: 'foo',
          type: 'action',
          exp: ['{{please}} = "work"'],
        },
        NodeType.Asset
      )
    ).toMatchSnapshot();
  });

  test('keeps null values when parsing object', () => {
    expect(parser.parseObject({ foo: 'bar', baz: null })).toStrictEqual({
      type: NodeType.Value,
      value: {
        foo: 'bar',
        baz: null,
      },
    });
  });
});

describe('parseView', () => {
  let model: DataModelWithParser;
  let expressionEvaluator: ExpressionEvaluator;
  let options: Options;
  let parser: Parser;

  beforeEach(() => {
    model = withParser(new LocalModel(), parseBinding);
    expressionEvaluator = new ExpressionEvaluator({
      model,
    });
    parser = new Parser();
    options = {
      evaluate: expressionEvaluator.evaluate,
      schema: new SchemaController(),
      data: {
        format: (binding, val) => val,
        formatValue: (val) => val,
        model,
      },
    };
    new TemplatePlugin(options).applyParser(parser);
    new ApplicabilityPlugin().applyParser(parser);
    new SwitchPlugin(options).applyParser(parser);
  });

  test('parses a simple view', () => {
    expect(
      parser.parseView({
        id: 'foo-view',
        type: 'viewtype',
        fields: {
          asset: {
            id: 'foo-asset',
            type: 'collection',
            values: [
              {
                asset: {
                  id: 'text-asset',
                  type: 'text',
                  value: 'bar',
                },
              },
              {
                asset: {
                  applicability: 'foo.bar',
                  id: 'input-asset',
                  type: 'input',
                },
              },
              {},
            ],
          },
        },
      } as any)
    ).toMatchSnapshot();
  });
});

describe('generates the correct AST when using switch plugin', () => {
  const toughStaticSwitchView = {
    id: 'toughView',
    type: 'view',
    title: {
      staticSwitch: [
        {
          case: "'true'",
          asset: {
            id: 'businessprofile-tile-screen-yoy-subtitle',
            type: 'text',
            value:
              "If it's changed since last year, let us know. Feel free to pick more than one.",
          },
        },
      ],
    },
  };

  const toughStaticSwitchMultiNodeView = {
    id: 'toughView',
    type: 'view',
    title: {
      asset: {
        id: 'someMultiNode',
        type: 'type',
        values: [
          {
            staticSwitch: [
              {
                case: "'true'",
                asset: {
                  id: 'businessprofile-tile-screen-yoy-subtitle-1',
                  type: 'text',
                  value:
                    "If it's changed since last year, let us know. Feel free to pick more than one.",
                },
              },
            ],
          },
          {
            asset: {
              id: 'businessprofile-tile-screen-yoy-subtitle-2',
              type: 'text',
              value: 'More text',
            },
          },
        ],
      },
    },
  };

  const switchPlugin = new SwitchPlugin({
    evaluate: () => {
      return true;
    },
  } as any);
  const parser = new Parser();
  switchPlugin.applyParser(parser);

  test('works with asset wrapped objects', () => {
    expect(parser.parseObject(toughStaticSwitchView)).toMatchSnapshot();
  });

  test('works with objects in a multiNode', () => {
    expect(
      parser.parseObject(toughStaticSwitchMultiNodeView)
    ).toMatchSnapshot();
  });
});
