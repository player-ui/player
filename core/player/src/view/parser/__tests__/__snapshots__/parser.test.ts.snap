// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates the correct AST parses an exp array 1`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "exp",
      ],
      "value": Object {
        "override": true,
        "parent": [Circular],
        "type": "multi-node",
        "values": Array [
          Object {
            "parent": [Circular],
            "type": "value",
            "value": "{{please}} = \\"work\\"",
          },
        ],
      },
    },
  ],
  "type": "asset",
  "value": Object {
    "id": "foo",
    "type": "action",
  },
}
`;

exports[`generates the correct AST parses an object 1`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "asset",
      ],
      "value": Object {
        "parent": [Circular],
        "type": "asset",
        "value": Object {
          "type": "bar",
        },
      },
    },
  ],
  "type": "value",
  "value": undefined,
}
`;

exports[`generates the correct AST when using switch plugin works with asset wrapped objects 1`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "title",
        "asset",
      ],
      "value": Object {
        "parent": [Circular],
        "type": "asset",
        "value": Object {
          "id": "businessprofile-tile-screen-yoy-subtitle",
          "type": "text",
          "value": "If it's changed since last year, let us know. Feel free to pick more than one.",
        },
      },
    },
  ],
  "type": "value",
  "value": Object {
    "id": "toughView",
    "type": "view",
  },
}
`;

exports[`generates the correct AST when using switch plugin works with objects in a multiNode 1`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "title",
        "asset",
      ],
      "value": Object {
        "children": Array [
          Object {
            "path": Array [
              "values",
            ],
            "value": Object {
              "override": true,
              "parent": [Circular],
              "type": "multi-node",
              "values": Array [
                Object {
                  "children": Array [
                    Object {
                      "path": Array [
                        "asset",
                      ],
                      "value": Object {
                        "parent": [Circular],
                        "type": "asset",
                        "value": Object {
                          "id": "businessprofile-tile-screen-yoy-subtitle-1",
                          "type": "text",
                          "value": "If it's changed since last year, let us know. Feel free to pick more than one.",
                        },
                      },
                    },
                  ],
                  "parent": [Circular],
                  "type": "value",
                  "value": undefined,
                },
                Object {
                  "children": Array [
                    Object {
                      "path": Array [
                        "asset",
                      ],
                      "value": Object {
                        "parent": [Circular],
                        "type": "asset",
                        "value": Object {
                          "id": "businessprofile-tile-screen-yoy-subtitle-2",
                          "type": "text",
                          "value": "More text",
                        },
                      },
                    },
                  ],
                  "parent": [Circular],
                  "type": "value",
                  "value": undefined,
                },
              ],
            },
          },
        ],
        "parent": [Circular],
        "type": "asset",
        "value": Object {
          "id": "someMultiNode",
          "type": "type",
        },
      },
    },
  ],
  "type": "value",
  "value": Object {
    "id": "toughView",
    "type": "view",
  },
}
`;

exports[`generates the correct AST works with applicability things 1`] = `
Object {
  "expression": "{{baz}}",
  "type": "applicability",
  "value": Object {
    "parent": [Circular],
    "type": "value",
    "value": Object {
      "foo": "bar",
    },
  },
}
`;

exports[`generates the correct AST works with applicability things 2`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "asset",
      ],
      "value": Object {
        "children": Array [
          Object {
            "path": Array [
              "values",
            ],
            "value": Object {
              "override": true,
              "parent": [Circular],
              "type": "multi-node",
              "values": Array [
                Object {
                  "expression": "{{foo}}",
                  "parent": [Circular],
                  "type": "applicability",
                  "value": Object {
                    "parent": [Circular],
                    "type": "value",
                    "value": Object {
                      "value": "foo",
                    },
                  },
                },
                Object {
                  "parent": [Circular],
                  "type": "value",
                  "value": Object {
                    "value": "bar",
                  },
                },
              ],
            },
          },
        ],
        "parent": [Circular],
        "type": "asset",
        "value": undefined,
      },
    },
  ],
  "type": "value",
  "value": undefined,
}
`;

exports[`generates the correct AST works with applicability things 3`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "asset",
      ],
      "value": Object {
        "children": Array [
          Object {
            "path": Array [
              "someProp",
            ],
            "value": Object {
              "expression": "{{foo}}",
              "parent": [Circular],
              "type": "applicability",
              "value": Object {
                "parent": [Circular],
                "type": "value",
                "value": Object {
                  "description": Object {
                    "value": "description",
                  },
                  "label": Object {
                    "value": "label",
                  },
                },
              },
            },
          },
        ],
        "parent": [Circular],
        "type": "asset",
        "value": undefined,
      },
    },
  ],
  "type": "value",
  "value": undefined,
}
`;

exports[`generates the correct AST works with applicability things 4`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "asset",
      ],
      "value": Object {
        "children": Array [
          Object {
            "path": Array [
              "someProp",
              "asset",
            ],
            "value": Object {
              "expression": "{{foo}}",
              "parent": [Circular],
              "type": "applicability",
              "value": Object {
                "parent": [Circular],
                "type": "asset",
                "value": Object {
                  "type": "someAsset",
                },
              },
            },
          },
        ],
        "parent": [Circular],
        "type": "asset",
        "value": undefined,
      },
    },
  ],
  "type": "value",
  "value": undefined,
}
`;

exports[`parseView parses a simple view 1`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "fields",
        "asset",
      ],
      "value": Object {
        "children": Array [
          Object {
            "path": Array [
              "values",
            ],
            "value": Object {
              "override": true,
              "parent": [Circular],
              "type": "multi-node",
              "values": Array [
                Object {
                  "children": Array [
                    Object {
                      "path": Array [
                        "asset",
                      ],
                      "value": Object {
                        "parent": [Circular],
                        "type": "asset",
                        "value": Object {
                          "id": "text-asset",
                          "type": "text",
                          "value": "bar",
                        },
                      },
                    },
                  ],
                  "parent": [Circular],
                  "type": "value",
                  "value": undefined,
                },
                Object {
                  "children": Array [
                    Object {
                      "path": Array [
                        "asset",
                      ],
                      "value": Object {
                        "expression": "foo.bar",
                        "parent": [Circular],
                        "type": "applicability",
                        "value": Object {
                          "parent": [Circular],
                          "type": "asset",
                          "value": Object {
                            "id": "input-asset",
                            "type": "input",
                          },
                        },
                      },
                    },
                  ],
                  "parent": [Circular],
                  "type": "value",
                  "value": undefined,
                },
              ],
            },
          },
        ],
        "parent": [Circular],
        "type": "asset",
        "value": Object {
          "id": "foo-asset",
          "type": "collection",
        },
      },
    },
  ],
  "type": "view",
  "value": Object {
    "id": "foo-view",
    "type": "viewtype",
  },
}
`;
